File: tst/buffer_raw.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    Pico_Rel_Dim phy = { '!', {100,100}, NULL };
    Pico_Rel_Dim log = { '!', { 10, 10}, NULL };
    pico_set_window("Buffer", -1, &phy);
    pico_set_view(-1, &log, NULL, NULL, NULL, NULL, NULL, NULL);

    // .x.
    // xxx
    // .x.
    Pico_Color_A buffer[] = {
        { 0x00, 0x00, 0x00, 0xFF },
        { 0xFF, 0xFF, 0x00, 0xFF },
        { 0x00, 0x00, 0x00, 0xFF },
        { 0xFF, 0x00, 0x00, 0xFF },
        { 0x00, 0xFF, 0x00, 0xFF },
        { 0x00, 0x00, 0xFF, 0xFF },
        PICO_COLOR_TRANSPARENT,
        { 0x00, 0xFF, 0xFF, 0xFF },
        PICO_COLOR_TRANSPARENT,
    };

    {
        puts("centered 3x3 on black - 1dir/1baixo");
        Pico_Rel_Rect r = { '!', {5-1,5-1,0,0}, PICO_ANCHOR_NW, NULL };
        pico_output_clear();
        pico_output_draw_buffer("buf1", (Pico_Abs_Dim){3,3}, buffer, &r);
        _pico_check("buffer-01");       // same as pct
    }

    {
        puts("bottomright 9x1 on white");
        pico_set_color_clear((Pico_Color){0xFF, 0xFF, 0xFF});
        pico_output_clear();
        Pico_Rel_Rect r = { '!', {1,9,0,0}, PICO_ANCHOR_NW, NULL };
        pico_output_draw_buffer("buf2", (Pico_Abs_Dim){9,1}, buffer, &r);
        _pico_check("buffer-02");       // same as pct
    }

    pico_init(0);
    return 0;
}

File: lua/tst/buffer_raw.lua
require 'pico.check'

pico.init(true)
local phy = {'!', w=100, h=100}
local log = {'!', w=10, h=10}
pico.set.window { title="Buffer", dim=phy }
pico.set.view { dim=log }

do
    -- .x.
    -- xxx
    -- .x.
    local buffer = {
        {
            { r=0x00, g=0x00, b=0x00 },             -- default a=0xFF
            { r=0xFF, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0xFF },
        },
        {
            { r=0xFF, g=0x00, b=0x00, a=0xFF },
            { r=0x00, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0xFF, a=0xFF },
        },
        {
            { r=0x00, g=0x00, b=0x00, a=0x00 },
            { r=0x00, g=0xFF, b=0xFF, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0x00 },
        },
    }

    print("centered 3x3 on black - 1dir/1baixo")
    local r = {'!', x=5-1, y=5-1, w=0, h=0, anc='NW'}
    pico.output.clear()
    pico.output.draw.buffer("buf1", buffer, r)
    pico.check("buffer-01")
end

do
    local buffer = {
        {
            { r=0x00, g=0x00, b=0x00, a=0xFF },
            { r=0xFF, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0xFF },
            { r=0xFF, g=0x00, b=0x00, a=0xFF },
            { r=0x00, g=0xFF, b=0x00, a=0xFF },
            { r=0x00, g=0x00, b=0xFF, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0x00 },
            { r=0x00, g=0xFF, b=0xFF, a=0xFF },
            { r=0x00, g=0x00, b=0x00, a=0x00 },
        }
    }

    print("bottomright 9x1 on white")
    pico.set.color.clear('white')
    pico.output.clear()
    local r = {'!', x=1, y=9, w=0, h=0, anc='NW'}
    pico.output.draw.buffer("buf2", buffer, r)
    pico.check("buffer-02")
end

pico.init(false)
